<xmemory> Add constexpr to allocator traits

MichaelRizkalla opened the pull request:
**Description**
---------------------------------
A small update that adds  `_CONSTEXPR20_DYNALLOC` to allocator traits member functions to implement constexpr allocators when the compiler defines `__cpp_constexpr_dynamic_alloc`.

This pull request implements the default allocator changes in [P0784R7](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0784r7.html) ~except for [allocator.traits.members] _construct_ and _destruct_ Effects~.

Useful for #37 

**Checklist**
---------------------------------
Before submitting a pull request, please ensure that:

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.

- [x] These changes introduce no known ABI breaks (adding members, renaming members, adding virtual functions, changing whether a type is an aggregate or trivially copyable, etc.).

- [x] Your changes are written from scratch using only this repository, the C++ Working Draft (including any cited standards), other WG21 papers (excluding reference implementations outside of proposed standard wording), and LWG issues as reference material. If your changes are derived from a project that's already listed in NOTICE.txt, that's fine, but please mention it. If your changes are derived from any other project, you *must* mention it here, so we can determine whether the license is compatible and what else needs to be done.

AdamBucior commented on stl/inc/xmemory:
```diff
@@ -606,7 +610,7 @@ struct _Normal_allocator_traits { // defines traits for allocators
 
 #if _HAS_IF_CONSTEXPR
-    _NODISCARD static size_type max_size(const _Alloc& _Al) noexcept {
+    _NODISCARD static _CONSTEXPR20_DYNALLOC size_type max_size(const _Alloc& _Al) noexcept {
```
This requires allocator::max_size() to be `constexpr` or removed. I don't know why it wasn't removed by #380.

AdamBucior replied:
I realised that it actually does not require it, because default allocator has it's own traits, but I still want to know why the deprecated `std::allocator` members weren't removed.

AdamBucior replied:
Also `_Default_allocator_traits`s members need to be made constexpr.

MichaelRizkalla replied:
I am not sure why allocator::max_size() is there yet.
However, if it is problematic, could it be made `constexpr` with `_CONSTEXPR20_DYNALLOC` macro?

miscco commented:
As @AdamBucior said, you need to add tests to `tests\std\tests\P0784R7_library_support_for_more_constexpr_containers\test.cpp`

Also as far as I know this would finish up P0784R7, so we should mark that in `<yvals_core>` as done

miscco commented:
Not I have a wip branch lying around that essentially sprinkels constexpr all around vector.

The problem that we have is that it is not clear currently what to do about `::operator new` and `::operator delete` if they take the alignment parameter, because those are at least in clang not `constexpr`

MichaelRizkalla commented on stl/inc/xmemory:
```diff
@@ -797,13 +801,15 @@ public:
+    _CONSTEXPR20_DYNALLOC allocator& operator=(const allocator&) = default;
 
-    void deallocate(_Ty* const _Ptr, const size_t _Count) {
+    _CONSTEXPR20_DYNALLOC void deallocate(_Ty* const _Ptr, const size_t _Count) {
```
`_Deallocate` was made `constexpr`.

MichaelRizkalla commented on stl/inc/xmemory:
```diff
@@ -797,13 +801,15 @@ public:
     }
 
-    _NODISCARD __declspec(allocator) _Ty* allocate(_CRT_GUARDOVERFLOW const size_t _Count) {
+    _NODISCARD _CONSTEXPR20_DYNALLOC __declspec(allocator) _Ty* allocate(_CRT_GUARDOVERFLOW const size_t _Count) {
```
`_Allocate` was made `constexpr`.

MichaelRizkalla commented:
> As @AdamBucior said, you need to add tests to `tests\std\tests\P0784R7_library_support_for_more_constexpr_containers\test.cpp`
> 
> Also as far as I know this would finish up P0784R7, so we should mark that in `<yvals_core>` as done

I'll add test coverage asap.

There are two missing changes for P0784R7 to be marked it as finished.
In [allocator.traits.members] paragraph 5 and 6:
allocator_trait::construct should invoke `construct_at(p, std::forward<Args>(args)...)` if the call is not well-formed
allocator_trait::destroy should invoke `destroy_at(p)`  if the call is not well-formed

`construct_at` and `destroy_at` are defined in `memory` header file in #501, so they would not be visible to allocator_traits.
It seemed reasonable to move `construct_at` and `destroy_at` to `xmemory` but since I am new here, I decided to avoid moving functions around until a complete dev guide comes out :)

miscco commented:
I am moving `construct_at` in https://github.com/microsoft/STL/pull/1329/commits/a3e3dc864a08258ca724c1f429b49bef24fbdea8

AdamBucior commented on tests/std/tests/VSO_0157762_feature_test_macros/test.cpp:
```diff
@@ -411,6 +411,20 @@ STATIC_ASSERT(__cpp_lib_constexpr_complex == 201711L);
 #endif
 #endif
 
+#if _HAS_CXX20
```
```suggestion
#if _HAS_CXX20 && defined(__cpp_constexpr_dynamic_alloc)
```

miscco commented on stl/inc/xmemory:
```diff
@@ -581,9 +587,14 @@ struct _Normal_allocator_traits { // defines traits for allocators
         if constexpr (_Uses_default_destroy<_Alloc, _Ty*>::value) {
+            (void) _Al;
+#if _HAS_CXX20 && defined(__cpp_constexpr_dynamic_alloc)
+            destroy_at(_Ptr)
```
```suggestion
            destroy_at(_Ptr);
```

mnatsuhara reviewed (changes requested):
Thanks for your work here!  It looks like there are some merge conflicts right now which is preventing the test run from kicking off.  I'll take another look once you have a chance to resolve the conflicts and address the review comments!

MichaelRizkalla commented on tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp:
```diff
@@ -272,6 +272,108 @@ constexpr void test_compiletime() {
+}
+static_assert((test_compiletime_allocator_traits(), true));
+
+constexpr void test_compiletime_allocator() {
```
Ops, I think I have missed the two operators. I'll add a test for them asap :).

MichaelRizkalla replied:
This should be safe to resolve now, the missing tests were added.

miscco commented on stl/inc/xmemory:
```diff
@@ -175,7 +176,7 @@ __declspec(allocator) void* _Allocate(const size_t _Bytes) {
+_CONSTEXPR20_DYNALLOC void _Deallocate(void* _Ptr, const size_t _Bytes) noexcept {
     // deallocate storage allocated by _Allocate when __cpp_aligned_new && _Align > __STDCPP_DEFAULT_NEW_ALIGNMENT__
     size_t _Passed_align = _Align;
 #if defined(_M_IX86) || defined(_M_X64)
```
This is not correct.

Changing the alignment here will not work. Please have a look at #1407 as I had to resolve the exact problem there

miscco commented on stl/inc/xmemory:
```diff
@@ -194,7 +195,7 @@ void _Deallocate(void* _Ptr, const size_t _Bytes) noexcept {
-__declspec(allocator) void* _Allocate(const size_t _Bytes) {
+__declspec(allocator) _CONSTEXPR20_DYNALLOC void* _Allocate(const size_t _Bytes) {
     // allocate _Bytes when !_HAS_ALIGNED_NEW || _Align <= __STDCPP_DEFAULT_NEW_ALIGNMENT__
 #if defined(_M_IX86) || defined(_M_X64)
```
Same here

miscco commented on stl/inc/xmemory:
```diff
@@ -210,7 +211,7 @@ __declspec(allocator) void* _Allocate(const size_t _Bytes) {
-void _Deallocate(void* _Ptr, size_t _Bytes) noexcept {
+_CONSTEXPR20_DYNALLOC void _Deallocate(void* _Ptr, size_t _Bytes) noexcept {
     // deallocate storage allocated by _Allocate when !_HAS_ALIGNED_NEW || _Align <= __STDCPP_DEFAULT_NEW_ALIGNMENT__
 #if defined(_M_IX86) || defined(_M_X64)
```
Ditto

miscco commented on stl/inc/xmemory:
```diff
@@ -547,16 +549,20 @@ struct _Normal_allocator_traits { // defines traits for allocators
+    static _CONSTEXPR20_DYNALLOC void construct(_Alloc& _Al, _Ty* _Ptr, _Types&&... _Args) {
         if constexpr (_Uses_default_construct<_Alloc, _Ty*, _Types...>::value) {
             (void) _Al; // TRANSITION, DevCom-1004719
+#if _HAS_CXX20 && defined(__cpp_constexpr_dynamic_alloc)
```
Please remove all occurrences of this pattern with `#ifdef __cpp_lib_constexpr_dynamic_alloc` (except for the one in yvals_core.h)

MichaelRizkalla replied:
Sorry for being that late, some vacation 😄. I will fix this pattern and the above reviews asap

miscco replied:
I would ask a maintainer, whether we want to replace the old functionality as done here or add an `if (_STD is_constant_evaluated())` branch

MichaelRizkalla replied:
The pattern has been replaced.
I will not resolve this one yet until a decision is made about using `if (_STD is_constant_evaluated())`

miscco reviewed (commented):
There are some correctness concerns regarding the special alignment changes we do

This suggests that there are insufficient tests there.

Also I am surprised that this works for sized deallocation, as that did not compile in ä1407 Cna you double check that these code paths are tested

Finaly, I am quite surprised that non of the libcxx tests are green now. Did you forget to mark them as passing for clang?

miscco commented on stl/inc/xmemory:
```diff
@@ -73,12 +73,13 @@ _INLINE_VAR constexpr size_t _New_alignof = (_STD max)(alignof(_Ty),
 // STRUCT _Default_allocate_traits
 struct _Default_allocate_traits {
-    __declspec(allocator) static void* _Allocate(const size_t _Bytes) {
+    __declspec(allocator) static _CONSTEXPR20_DYNALLOC void* _Allocate(const size_t _Bytes) {
```
As shown in the commits on my subbranch, this can (but must) be constexpr for clang, as MSVC only allows std::allocator<T>::allocate

MichaelRizkalla replied:
Resolved from 35e4aeb

miscco commented on stl/inc/xmemory:
```diff
@@ -73,12 +73,13 @@ _INLINE_VAR constexpr size_t _New_alignof = (_STD max)(alignof(_Ty),
 
 #ifdef __cpp_aligned_new
-    __declspec(allocator) static void* _Allocate_aligned(const size_t _Bytes, const size_t _Align) {
+    __declspec(allocator) static _CONSTEXPR20_DYNALLOC
```
Same here, the _CONSTEXPR20_DYNALLOC must also be guarded by `#ifdef __clang__`

miscco commented on stl/inc/xmemory:
```diff
@@ -73,12 +73,13 @@ _INLINE_VAR constexpr size_t _New_alignof = (_STD max)(alignof(_Ty),
 #ifdef __cpp_aligned_new
-    __declspec(allocator) static void* _Allocate_aligned(const size_t _Bytes, const size_t _Align) {
+    __declspec(allocator) static _CONSTEXPR20_DYNALLOC
+        void* _Allocate_aligned(const size_t _Bytes, const size_t _Align) {
```
I think I added it to the commit, but for some reason aligned allocation do not currently work with clang so this should fall back to unaligned `::operator new(_Ptr);` during constant evaluation

miscco commented on stl/inc/xmemory:
```diff
@@ -675,35 +686,36 @@ struct _Default_allocator_traits { // traits for std::allocator
     using rebind_traits = allocator_traits<allocator<_Other>>;
 
-    _NODISCARD static __declspec(allocator) pointer allocate(_Alloc&, _CRT_GUARDOVERFLOW const size_type _Count) {
+    _NODISCARD static _CONSTEXPR20_DYNALLOC __declspec(allocator) pointer
```
So this needs some love for MSVC, as only `std::allocator::allocate` is allowed during constant evalaution:

```cpp    
    _NODISCARD static _CONSTEXPR20_DYNALLOC __declspec(allocator) pointer
        allocate(_Alloc& _Al, _CRT_GUARDOVERFLOW const size_type _Count) {
#ifdef __cpp_lib_is_constant_evaluated
        if (_STD is_constant_evaluated()) {
            return _Al.allocate(_Count);
        } else
#endif // __cpp_lib_is_constant_evaluated
        {
            (void) _Al;
            return static_cast<pointer>(
                _Allocate<_New_alignof<value_type>>(_Get_size_of_n<sizeof(value_type)>(_Count)));
        }
    }
```

This also applies for the other functions below

miscco commented on stl/inc/xmemory:
```diff
@@ -675,35 +686,36 @@ struct _Default_allocator_traits { // traits for std::allocator
 
     template <class _Objty, class... _Types>
-    static void construct(_Alloc&, _Objty* const _Ptr, _Types&&... _Args) {
+    static _CONSTEXPR20_DYNALLOC void construct(_Alloc&, _Objty* const _Ptr, _Types&&... _Args) {
```
Here we need to add `_STD construct_at(_Ptr, _STD forward<_Types>(_Args)...)` during constant evaluation

miscco commented on stl/inc/xmemory:
```diff
@@ -675,35 +686,36 @@ struct _Default_allocator_traits { // traits for std::allocator
     template <class _Objty, class... _Types>
-    static void construct(_Alloc&, _Objty* const _Ptr, _Types&&... _Args) {
+    static _CONSTEXPR20_DYNALLOC void construct(_Alloc&, _Objty* const _Ptr, _Types&&... _Args) {
         ::new (const_cast<void*>(static_cast<const volatile void*>(_Ptr))) _Objty(_STD forward<_Types>(_Args)...);
```
Can we move this to `_Voidify`?

MichaelRizkalla replied:
I think it can be moved.

miscco commented on stl/inc/xmemory:
```diff
@@ -675,35 +686,36 @@ struct _Default_allocator_traits { // traits for std::allocator
 
     template <class _Uty>
-    static void destroy(_Alloc&, _Uty* const _Ptr) {
+    static _CONSTEXPR20_DYNALLOC void destroy(_Alloc&, _Uty* const _Ptr) {
```
This needs `_STD destroy_at(_Ptr)` during constant evaluation

miscco commented on tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp:
```diff
@@ -272,8 +272,123 @@ constexpr void test_compiletime() {
+    constexpr size_type max_size() noexcept {
+        return std::numeric_limits<size_type>::max() / sizeof(value_type);
+    }
+
```
I believe we should also have test for an allocator with/without a `construct` and `destroy` methods, to ensure that they are properly forwarded.

I guess two additional bool template arguments and some SFINAE magic are the least pain here

MichaelRizkalla replied:
I added 3 more cases with/without `construct` and `destroy` methods.

miscco reviewed (commented):
I am sorry for bombarding you with comments :(

miscco commented on stl/inc/xmemory:
```diff
@@ -675,35 +739,75 @@ struct _Default_allocator_traits { // traits for std::allocator
+        } else
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            return static_cast<pointer>(
```
This will fail because of unused `_Al`
```suggestion
            (void) _Al;
            return static_cast<pointer>(
```

miscco reviewed (approved):
Last comment, but after resolving what the maintainers prefer for changing the allocators this looks great

mnatsuhara commented:
@MichaelRizkalla - I've pushed a few changes to your branch in the interest of preparing this for review this evening 😸  Summary of changes:
- Clang 11 formatting changes in `<xmemory>`, `<xutility>`
- Compilation errors:
   - `#endif`s should be commented with the condition of its paired `#if` or `#ifdef`, not be followed by the condition itself (e.g., `#endif _HAS_CXX20` --> `#endif // _HAS_CXX20`) 
      - Changes made in `<xmemory>`, `<xutility>`
   - MSVC has some blocking bugs (in particular, issues with `constexpr` destructors) that prevent us from enabling full functionality of P0784, so I've adjusted the definition of `__cpp_lib_constexpr_dynamic_alloc` and `_CONSTEXPR20_DYNALLOC` definition in `<yvals_core.h>` to only expand to `constexpr` for Clang and EDG (consistent with #1502 and #1407 once it has `defined(__EDG__)` added). 
   - `enable_if_t` wasn't usable to conditionally define `Alloc::construct` and `Alloc::destroy` (I believe this is because there needs to be _some_ overload that is selected, it can't just determine whether a member function exists at all) so I changed those functions to use concepts, which is similar to the strategy used in some of the concepts/ranges tests.  e.g.: https://github.com/microsoft/STL/blob/5be862db893972a95d2f8b223e86a4a8e8f24839/tests/std/tests/P0896R4_P1614R2_comparisons/test.cpp#L79-L101
- Getting the test to pass:
   - Because `std::allocator::construct` and other functions are deprecated in cxx17 and beyond, their use in the user-defined `Alloc` in the tests was emitting the deprecation warnings.  I marked those deprecated functions with `_CONSTEXPR20_DYNALLOC` and `#define`d the `_SILENCE_CXX17_OLD_ALLOCATOR_MEMBERS_DEPRECATION_WARNING` in the test for now, though I don't know if this is the best fix overall. 
   - There was an issue of calling the destructor of an object whose lifetime had already ended, given the destructor of `nontrivial_A` - since I believe a nontrivial destructor just needs to be defined, I removed the call to the destructor and just left the body empty. 
   - Only enable the added tests for Clang and EDG (consistent with the adjusted conditions for `_CONSTEXPR20_DYNALLOC` and `__cpp_lib_constexpr_dynamic_alloc` as mentioned above). 
   - `alloc` was referenced in some blocks without anything in scope with that name, so I added `Alloc` variable definitions for use there. 
- Minor cleanup: Removed unnecessary semicolons at the ends of method definitions in `P0784_library_support_for_more_constexpr_containers/test.cpp`, which also caused clang-format to adjust the spacing of the `{}`

MichaelRizkalla commented on tests/std/tests/VSO_0157762_feature_test_macros/test.compile.pass.cpp:
```diff
@@ -411,6 +411,20 @@ STATIC_ASSERT(__cpp_lib_constexpr_complex == 201711L);
 #endif
 #endif
 
+#if _HAS_CXX20 && defined(__cpp_constexpr_dynamic_alloc)
```
```suggestion
#if _HAS_CXX20 && defined(__cpp_constexpr_dynamic_alloc)  && (defined(__clang__) || defined(__EDG__))
```

@mnatsuhara should we add the constraints to the feature test macro as well since `yval_cores.h` will not define `__cpp_lib_constexpr_dynamic_alloc` except for Clang and EDG?

mnatsuhara replied:
Yeah, I realized a bit too late that I forgot to update this as well :)

mnatsuhara commented on tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp:
```diff
@@ -272,8 +276,152 @@ constexpr void test_compiletime() {
+#endif // defined(__clang__ ) || defined(__EDG__)
 #endif // _HAS_CXX20 && defined(__cpp_constexpr_dynamic_alloc)
 
+#if _HAS_CXX20
```
We don't need any switches for `_HAS_CXX20` because P0784 is C++20-specific and this test is run with `usual_latest_matrix` which always implies C++20 (or beyond) mode!  There are some `_HAS_CXX20`s pre-existing in this file, so please audit the entire test to remove these.

mnatsuhara reviewed (changes requested):
I did a full review of product code and test coverage.  This is looking really good!  There are a few comments, etc. that were missed (related to our tracking of progress / infrastructure) and a few other smaller issues, but overall your changes look great!  Thanks so much for your work here!

miscco commented on stl/inc/xmemory:
```diff
@@ -677,35 +741,76 @@ struct _Default_allocator_traits { // traits for std::allocator
-        return static_cast<pointer>(_Allocate<_New_alignof<value_type>>(_Get_size_of_n<sizeof(value_type)>(_Count)));
+    _NODISCARD static _CONSTEXPR20_DYNALLOC __declspec(allocator) pointer
+        allocate(_Alloc& _Al, _CRT_GUARDOVERFLOW const size_type _Count) {
+#ifdef __cpp_lib_is_constant_evaluated
+        if (_STD is_constant_evaluated()) {
+            return _Al.allocate(_Count);
+        } else
+#endif // __cpp_lib_is_constant_evaluated
```
I believe I was the culprit here.

From my understanding there is a difference between MSCV and clang.

MSVC will replace `_Al.allocate` with fairy dust, as this is `allocator::allocate`
clang will go into `_Allocate`

mnatsuhara replied:
This comment was intended to point out that we have escape hatches for `constexpr` things in both this function and in the `_Allocate` function that we are evidently avoiding if `std::allocator::allocate()` is called in a `constexpr` context.  It seems like it would be nice to centralize the logic for workarounds for `constexpr` things in one area and sort of determine whether `_Allocate` is completely safe for `constexpr` or not.  This is definitely not something that needs to be included in this PR, but I want to file an issue for it to at least take another look to see if we can make this dance a little nicer given the approaches from Clang and MSVC.  

From my understanding: 
MSVC does fairy magic when it sees `std::allocator::allocate`, but doesn't otherwise allow `::operator new` so we need to go through `std::allocator::allocate` in `allocator_traits::allocate`
Clang allows calls to `::operator new` if they are reached transitively from an evaluation of `std::allocator::allocate` so we can maintain the call to `_Allocate` directly, or go through `std::allocator::allocate` which does a call to `_Allocate`.

mnatsuhara replied:
Follow up here: I've filed GH-1532 to track this so the transition comment can read `// TRANSITION, GH-1532`.

miscco commented on tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp:
```diff
@@ -272,8 +276,152 @@ constexpr void test_compiletime() {
+        allocator<T>{}.deallocate(ptr, n);
+    }
+
+    constexpr void construct(value_type* ptr, value_type n) requires(Construct) {
+        allocator<T>{}.construct(ptr, n);
+    }
+
+    constexpr void destroy(value_type* ptr) requires(Destroy) {
+        allocator<T>{}.destroy(ptr);
+    }
```
I am not sure whether this is correct, we never really remove something so we should test for someone pulling the escape hatch too

mnatsuhara replied:
Pulling the escape hatch as in `_Alloc.construct` / `_Alloc.destroy` not being available, and needing to call `construct_at` and `destroy_at` instead?  I believe the `requires` should conditionally remove `construct` or `destroy` from `Alloc`'s member functions, so that is how we are exercising this escape hatch.  

Getting into the details of the escape hatch and corresponding changes: 
After consulting with @StephanTLavavej and @CaseyCarter, I don't think we have a nice way to verify that we are calling `std::construct_at` / `std::destroy_at` instead of `::new` and `~_Ty()` (one of the changes in WG21-P0784).  Just seeing the change in the product code is sufficient and consistent with similar changes elsewhere, at least for the time being.

miscco commented on stl/inc/xmemory:
```diff
@@ -799,32 +904,34 @@ public:
     constexpr allocator(const allocator&) noexcept = default;
     template <class _Other>
     constexpr allocator(const allocator<_Other>&) noexcept {}
+    _CONSTEXPR20_DYNALLOC ~allocator()       = default;
+    _CONSTEXPR20_DYNALLOC allocator& operator=(const allocator&) = default;
 
-    void deallocate(_Ty* const _Ptr, const size_t _Count) {
+    _CONSTEXPR20_DYNALLOC void deallocate(_Ty* const _Ptr, const size_t _Count) {
         // no overflow check on the following multiply; we assume _Allocate did that check
         _Deallocate<_New_alignof<_Ty>>(_Ptr, sizeof(_Ty) * _Count);
     }
 
-    _NODISCARD __declspec(allocator) _Ty* allocate(_CRT_GUARDOVERFLOW const size_t _Count) {
+    _NODISCARD _CONSTEXPR20_DYNALLOC __declspec(allocator) _Ty* allocate(_CRT_GUARDOVERFLOW const size_t _Count) {
         return static_cast<_Ty*>(_Allocate<_New_alignof<_Ty>>(_Get_size_of_n<sizeof(_Ty)>(_Count)));
     }
 
-    _CXX17_DEPRECATE_OLD_ALLOCATOR_MEMBERS _NODISCARD __declspec(allocator) _Ty* allocate(
+    _CXX17_DEPRECATE_OLD_ALLOCATOR_MEMBERS _NODISCARD _CONSTEXPR20_DYNALLOC __declspec(allocator) _Ty* allocate(
```
Same here, I believe some users will keep them and we will have to support it

MichaelRizkalla commented on stl/inc/memory:
```diff
@@ -567,22 +567,7 @@ namespace ranges {
     };
 
```
I can add these changes and the test coverage for them. But I have to ask, should the parallel version be included? I can not see any implementation of them 🐱

mnatsuhara replied:
Awesome!  Thanks 😸 

Adding `constexpr` to the parallel versions of `destroy`, `destroy_n` in WG21-P0784 was a mistake that was corrected in LWG-3303.  We shouldn't need to `constexpr`ize the parallel versions!  Threading / Concurrency is essentially "not a thing" in `constexpr` land (yet 😉).

MichaelRizkalla commented on stl/inc/xmemory:
```diff
@@ -73,13 +73,28 @@ _INLINE_VAR constexpr size_t _New_alignof = (_STD max)(alignof(_Ty),
 
 // STRUCT _Default_allocate_traits
 struct _Default_allocate_traits {
+#ifdef __clang__
```
I am not familiar with `// TRANSITION, ...` comments. Should it include a specific PR, issue, or something else?

cbezault replied:
If there is a specific issue number/PR either in LLVM land or here in this repo you can put it there. If the reason is relatively simple you can just write it out.

MichaelRizkalla commented on stl/inc/xutility:
```diff
@@ -132,10 +132,27 @@ _NODISCARD void* _Voidify_iter(_Iter _It) noexcept {
+#if _HAS_CXX20
+template <class _Ty, class... _Types>
+_CONSTEXPR20_DYNALLOC auto construct_at(_Ty* const _Location, _Types&&... _Args) noexcept(
+    noexcept(::new (_Voidify_iter(_Location)) _Ty(_STD forward<_Types>(_Args)...))) // strengthened
+    -> decltype(::new (_Voidify_iter(_Location)) _Ty(_STD forward<_Types>(_Args)...)) {
+    return ::new (_Voidify_iter(_Location)) _Ty(_STD forward<_Types>(_Args)...);
+}
+#endif // _HAS_CXX20
+
 // FUNCTION TEMPLATE _Construct_in_place
 template <class _Ty, class... _Types>
 void _Construct_in_place(_Ty& _Obj, _Types&&... _Args) noexcept(is_nothrow_constructible_v<_Ty, _Types...>) {
-    ::new (_Voidify_iter(_STD addressof(_Obj))) _Ty(_STD forward<_Types>(_Args)...);
+#ifdef __cpp_lib_is_constant_evaluated
+    if (_STD is_constant_evaluated()) {
+        _STD construct_at(_STD addressof(_Obj), _STD forward<_Types>(_Args)...);
+    } else
+#endif // __cpp_lib_is_constant_evaluated
+    {
+        ::new (_Voidify_iter(_STD addressof(_Obj))) _Ty(_STD forward<_Types>(_Args)...);
+    }
```
Indeed, it was borrowed. I'll revert it back.

MichaelRizkalla commented:
@mnatsuhara, I have updated the PR with some of the review comments, and I have some notes 🐱 

- using `_CONSTEXPR20_DYNALLOC` instead of `_CONSTEXPR20` with `max_size` and `select_on_container_copy_construction` in `_Default_allocator_traits` and `_Normal_allocator_traits` will result in some failing tests. Same for `operator==` and `operator!=`. I suppose that is because `_CONSTEXPR20_DYNALLOC` is interpreted as `inline` with the current setup, right? 🤔
- I fixed a small bug, at least I think it is a bug 😼, in `_Destroy_n_fn::operator()` where a const variable was decremented. Can you double-check if that's correct?
- for the file `P0784R7_library_support_for_more_constexpr_containers/test.cpp`:
  - The file is cleaned from `_HAS_CXX20`, and using `__cpp_lib_constexpr_dynamic_alloc` instead
  - Couple of new tests for `destroy`, `destroy_n`, `ranges::destroy` and `ranges::destroy_n` were added

mnatsuhara reviewed (changes requested):
Thanks for implementing the changes requested!  A few more follow up things with the second pass through 🐱 The 🐞 fix for decrementing a `const` variable looks good to me, and I think I've answered the other questions through my comments!

MichaelRizkalla commented on tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp:
```diff
@@ -272,7 +266,208 @@ constexpr void test_compiletime() {
+        }
+        destroy_n(begin(a), 10);
+    }
+#ifdef __cpp_lib_concepts
+    {
+        A<int> a[10];
+        for (int i = 0; i < 10; i++) {
+            ranges::construct_at(&a[i].value, i);
+        }
+        ranges::destroy_n(ranges::begin(a), 10);
+    }
+#endif // __cpp_lib_concepts
```
Yes, I intended to use the `nontrivial_A<int>`

MichaelRizkalla commented on stl/inc/xmemory:
```diff
@@ -677,35 +741,76 @@ struct _Default_allocator_traits { // traits for std::allocator
-    static void destroy(_Alloc&, _Uty* const _Ptr) {
-        _Ptr->~_Uty();
+    static _CONSTEXPR20_DYNALLOC void destroy(_Alloc&, _Uty* const _Ptr) {
+#ifdef __cpp_lib_is_constant_evaluated
+        if (_STD is_constant_evaluated()) {
+            _STD destroy_at(_Ptr);
+        } else
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            _Ptr->~_Uty();
+        }
```
I am in favor of using `_STD destroy_at` unconditionally to avoid code duplicates. I'll add that next commit.

MichaelRizkalla replied:
More follow up:
I am thinking of an implementation like that could work as well
```C++
#if _HAS_CXX17
    _STD destroy_at(_Ptr);
#else
    _Ptr->~Uty();
#endif
```
What do you think?

mnatsuhara replied:
From my reading, we are considering 2 different things that we can toggle on/off (whether the function is constant evaluated, and whether this function is called in C++20 mode); this results in 4 scenarios:

1. C++20 and constant-evaluated
2. C++20 and not constant-evaluated
3. pre-C++20 and constant-evaluated
4. pre-C++20 and not constant-evaluated

Option 3 should not be possible, as this function is not marked `constexpr` until C++20.  

Options 1 and 2 should call `_STD destroy_at` given WG21-P0784, since change in the description of `std::allocator_traits::destroy` does not hinge on whether the function is constant-evaluated or not, it simply changes that behavior in C++20.  

Option 4 should preserve pre-C++20 behavior and continue to call `_Ptr->~Uty()` as it did before.  (I was mistaken in my original comment - we can't call `_STD destroy_at` _unconditionally_, but I believe we should do it in more cases than this current implementation does).  I think we can utilize a pattern similar to what you have in your follow up comment to do this: 
```C++
#ifdef __cpp_lib_constexpr_dynamic_alloc
   _STD destroy_at(_Ptr);
#else
   _Ptr->~Uty();
#endif // __cpp_lib_constexpr_dynamic_alloc
```

MichaelRizkalla replied:
This should be resolved with commit 686b519 and 580db38

mnatsuhara reviewed (changes requested):
I just have one small note on test coverage, but otherwise this looks great!  Once we resolve the comment about `_Default_allocator_traits::destroy`, I think this will be ready to move on to a second reviewer!

MichaelRizkalla commented on tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp:
```diff
@@ -272,7 +266,229 @@ constexpr void test_compiletime() {
+        for (int i = 0; i < 10; i++) {
+            ranges::construct_at(a + i);
+        }
+        ranges::destroy(a, a + 10);
```
I was happy that the tests finally passed that I forgot to check if I covered everything 😹, my bad.

miscco commented on stl/inc/xmemory:
```diff
@@ -73,13 +73,28 @@ _INLINE_VAR constexpr size_t _New_alignof = (_STD max)(alignof(_Ty),
     }
 
 #ifdef __cpp_aligned_new
+#ifdef __clang__ // TRANSITION, MSVC support for constexpr dynamic allocation
```
The transition comment  below is different than this one, but it describes the same feature request for MSVC, so we should use the same comment here and abve

miscco commented on stl/inc/xmemory:
```diff
@@ -677,35 +740,73 @@ struct _Default_allocator_traits { // traits for std::allocator
-        return static_cast<pointer>(_Allocate<_New_alignof<value_type>>(_Get_size_of_n<sizeof(value_type)>(_Count)));
+    _NODISCARD static _CONSTEXPR20_DYNALLOC __declspec(allocator) pointer
+        allocate(_Alloc& _Al, _CRT_GUARDOVERFLOW const size_type _Count) {
+#ifdef __cpp_lib_is_constant_evaluated // TRANSITION, GH-1532
```
I am super sorry, I think you took that one from me. 

Above you guard destruction with `__cpp_lib_constexpr_dynamic_alloc`. I *think* this would be the right choice here too, as otherwise we do not need that code at all.

Please wait for a maintainers verdict before you go along and change all of those

MichaelRizkalla replied:
I think this change can also be done to both `allocate` overloads, `deallocate`, and `construct` below. Waiting for a maintainer's decision 🐱.

mnatsuhara replied:
If I'm understanding correctly, you are proposing to change the `#ifdef __cpp_lib_is_constant_evaluated` to `#ifdef __cpp_lib_constexpr_dynamic_alloc`?  I think we are in an interesting position where they function the same way here, so I don't know if it matters hugely.  

< (This is just me trying to reason through this, feel free to skip to the last paragraph if you already agree) >
The new behavior of potentially going through `_Al.(de)allocate` in `allocate` and `deallocate` only really needs to be applied if the functions are constant evaluated.  Because WG21-P0784 is the paper making the functions `constexpr`, if `__cpp_lib_is_constant_evaluated()` is defined and `_STD is_constant_evaluated() == true`, then we know `__cpp_lib_constexpr_dynamic_alloc` is defined as well.  If the function is not constant evaluated (whether that is because WG21-P0784 hasn't been implemented or just because it isn't called at compile-time), the old behavior should remain.  

This same logic applies to the change in `construct`.  We could just toggle `construct_at` vs. `::new` on `__cpp_lib_constexpr_dynamic_alloc` as we do in `destroy`, but here we eliminate a function call if the call is not made in `constexpr` land (the body of `construct_at` is the same as the pre-existing `::new` call).  Note that `destroy_at` actually has more logic than just calling a destructor (which is the pre-existing body of the function), which is why we just toggle based on `__cpp_lib_constexpr_dynamic_alloc` there.  

If we change to using `__cpp_lib_constexpr_dynamic_alloc` instead of `__cpp_lib_is_constant_evaluated`, we know that `_STD is_constant_evaluated()` is available, so there won't be any behavioral change there.  

In other words, I believe we would get the same behavior either way, so I think this question is reduced to a matter of convention.  I'm not aware of a strict convention here, but looking through the code base, I see primarily uses of `__cpp_lib_is_constant_evaluated` and `__cpp_lib_concepts`, and not many other feature test macros used.  This indicates that maybe we should stick with `__cpp_lib_is_constant_evaluated`, though I yield to another maintainer who feels more strongly.

miscco replied:
So my reasoning is slightly different and given that all supported compilers either support both or nothing it is kind of moot.

The point is that `__cpp_lib_constexpr_dynamic_alloc` is a stronger constraint than `__cpp_lib_is_constant_evaluated` in the sense that all compilers first implemented `__cpp_lib_is_constant_evaluated` and then `__cpp_lib_constexpr_dynamic_alloc`

So right now we are in the situation that a compiler that supports only `__cpp_lib_is_constant_evaluated` (MSVC right now) has to parse/validate those code snippets, although he can never use them. However, if we switch it around, only compilers that actually support `__cpp_lib_constexpr_dynamic_alloc` will see the code that is needed for it.

Does that make sense?

mnatsuhara replied:
Yeah, that makes sense.  I don't feel strongly either way - I am fine changing to `__cpp_lib_constexpr_dynamic_alloc` and whoever the second reviewer is can chime in if they strongly disagree :)

MichaelRizkalla replied:
I will change it to `__cpp_lib_constexpr_dynamic_alloc`  just from a consistency point of view in all affected member functions of `_Default_allocator_traits`. Hopefully, this will be the last change in this PR 😂.

miscco commented on stl/inc/yvals_core.h:
```diff
@@ -554,7 +555,8 @@
 // Functions that became constexpr in C++20 via P0784R7
-#if _HAS_CXX20 && defined(__cpp_constexpr_dynamic_alloc)
+#if _HAS_CXX20 && defined(__cpp_constexpr_dynamic_alloc) \
+    && (defined(__clang__) || defined(__EDG__)) // TRANSITION, MSVC support for constexpr dynamic allocation
```
All thisconstexpr stuf makes my head spin. As far as I know, we only have the constexpr destructor test that fails with MSVC.

What happens when we define this unconditionally?

MichaelRizkalla replied:
As far as I understand, this was applied because of the constexpr destructors only.
If defined unconditionally, most of the test cases in `P0784R7_library_support_for_more_constexpr_containers` will fail for MSVC because they rely on the constexpr destructors.

mnatsuhara replied:
Since we can't enable full `_CONSTEXPR20_DYNALLOC` for MSVC for `string`/`vector` due to the `constexpr` destructor issue, the idea was to just consider MSVC as not supporting `constexpr` dynamic allocation just yet and turn everything on at the same time.

miscco replied:
Thanks, I forgot about the destructor bug

miscco commented on tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp:
```diff
@@ -272,7 +266,249 @@ constexpr void test_compiletime() {
+    constexpr ~nontrivial_A() {}
+};
+
+constexpr void test_compiletime_destroy_variants() {
```
The one difference that we have between the old `destroy` and the new version is that we utilize `destroy_at` which handles arrays special.

Could we add some test coverage for something that stores an array of ints?

MichaelRizkalla replied:
I believe there is already a test that covers the array path in `destroy_at` and `ranges::destroy_at` in here:
https://github.com/microsoft/STL/blob/cf86b05c5e86e8a91cc67a57e622c55d06414091/tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp#L205-L213

miscco reviewed (commented):
Thanks a lot for keeping the ball rolling. This is really tricky stuff with lots of sharp edges.

I have some comments but this is already looking really solid :+1:

mnatsuhara reviewed (commented):
Just a small nitpick in `test.cpp` and a question of consistency given the recent change with feature-test macros 🐱

MichaelRizkalla commented on stl/inc/xmemory:
```diff
@@ -73,13 +73,28 @@ _INLINE_VAR constexpr size_t _New_alignof = (_STD max)(alignof(_Ty),
     __declspec(allocator) static void* _Allocate_aligned(const size_t _Bytes, const size_t _Align) {
-        return ::operator new (_Bytes, align_val_t{_Align});
+#ifdef __clang__ // TRANSITION, MSVC support for constexpr dynamic allocation
+#ifdef __cpp_lib_is_constant_evaluated
```
I overlooked these occurrences, fixed now!

mnatsuhara reviewed (approved):
Awesome, thanks for all of your work and follow up here!  This is now ready for final review!

mnatsuhara commented:
FYI @MichaelRizkalla, I merged your branch with master to resolve the merge conflicts after #1544 was checked in!

StephanTLavavej commented on tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp:
```diff
@@ -272,7 +267,249 @@ constexpr void test_compiletime() {
+        return Alloc{id + 1};
+    }
+
+    constexpr size_type max_size() noexcept {
```
This can be `const`.

StephanTLavavej commented on tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp:
```diff
@@ -272,7 +267,249 @@ constexpr void test_compiletime() {
+}
+static_assert((test_compiletime_destroy_variants(), true));
+
+template <class T, bool Construct = false, bool Destroy = false>
+struct Alloc {
```
This allocator is missing:
* A rebinding constructor
* `operator==` (C++20 will rewrite `operator!=`)
* `rebind` support

For `rebind`, there's a default for allocators of the form `SomeAllocator<T, Args>` (see WG21-N4878 16.4.4.6.1 [allocator.requirements.general]/3), but the `bool` template parameters interfere with this.

StephanTLavavej commented on tests/std/tests/P0784R7_library_support_for_more_constexpr_containers/test.cpp:
```diff
@@ -272,7 +267,249 @@ constexpr void test_compiletime() {
+constexpr void test_compiletime_allocator_traits() {
+    {
+        storage_for<A<int>> a;
+        Alloc<A<int>> alloc{10};
+        assert(alloc.id == 10);
+
+        auto result = std::allocator_traits<Alloc<A<int>>>::allocate(alloc, 10);
```
Observation, no change requested: For clarity, I recommend using distinct values for distinct purposes in test code. Here, `10` is being used for the allocator's ID, and `10` is also the number of `A<int>` elements being allocated. This makes it more difficult to glance at the source code and immediately recognize the meaning of a `10`.

StephanTLavavej reviewed (commented):
Thanks, this looks good! I had a number of comments but they're all straightforward to resolve; in the interests of time I will prepare a commit to address them.

StephanTLavavej commented on stl/inc/xmemory:
```diff
@@ -529,38 +583,47 @@ struct _Normal_allocator_traits { // defines traits for allocators
-    static void destroy(_Alloc& _Al, _Ty* _Ptr) {
+    static _CONSTEXPR20_DYNALLOC void destroy(_Alloc& _Al, _Ty* _Ptr) {
         if constexpr (_Uses_default_destroy<_Alloc, _Ty*>::value) {
+            (void) _Al;
```
It appears that this occurrence in `destroy` is unnecessary, because there's an `if constexpr` branch using `_Al` - however, the occurrence in `deallocate` is necessary because the usage of `_Al` is guarded with `#ifdef __cpp_lib_constexpr_dynamic_alloc`. So that suppression should be preserved, without a `TRANSITION`. Checking all occurrences.

StephanTLavavej replied:
This was the only unnecessary occurrence, the others were all necessary.

StephanTLavavej commented on stl/inc/memory:
```diff
@@ -661,8 +629,9 @@ namespace ranges {
-        /* _CONSTEXPR20_DYNALLOC */ _It operator()(_It _First, const iter_difference_t<_It> _Count) const noexcept {
+         _CONSTEXPR20_DYNALLOC _It operator()(_It _First, const iter_difference_t<_It> _Count_raw) const noexcept {
             // clang-format on
+            _Algorithm_int_t<iter_difference_t<_It>> _Count = _Count_raw;
```
Nope, I was wrong! There was a pre-existing bug here! Previously, this code had a `const iter_difference_t<_It> _Count` parameter, but below it says `--_Count;`. I have *no idea* why the tests were passing before. Your change is correct.

StephanTLavavej commented:
@MichaelRizkalla @mnatsuhara @miscco I believe this is ready to merge, but if you see any issues with the changes I just pushed (particularly the compiler guard changes), we can continue iterating.

miscco commented on stl/inc/xmemory:
```diff
@@ -73,27 +73,18 @@ _INLINE_VAR constexpr size_t _New_alignof = (_STD max)(alignof(_Ty),
-        _CONSTEXPR20_DYNALLOC
-#endif // __clang__
-        void* _Allocate(const size_t _Bytes) {
+    __declspec(allocator) static _CONSTEXPR20_DYNALLOC void* _Allocate(const size_t _Bytes) {
```
This is not correct, MSVC does no accept operator new in a constant expression, but only `construct_at`.

This is essentially a clang extension

miscco replied:
Note this only compiles now because `_CONSTEXPR20_DYNALLOC` gets sustituted to `inline` for MSVC

MichaelRizkalla replied:
It seems we are allowing `::operator delete` in `_Deallocate` to be invoked from constexpr context, does your comment apply to it as well?.

miscco replied:
That is a good question, we should ping cadacama

mnatsuhara replied:
Let's bring @StephanTLavavej back into the conversation, since this was his change 🐱 
To clarify, only enabling `_CONSTEXPR20_DYNALLOC` for Clang here is because MSVC and Clang have fundamentally different approaches in how they are implementing WG21-P0784: 

- Clang allows `::operator new` in `constexpr` contexts when reached transitively from a call to `std::allocator::allocate`.
- MSVC does fairy magic with `std::allocator::allocate` behind the scenes to allow it, but does not allow `::operator new` in `constexpr` contexts.  

(a bit more discussion of this in #1532)

So, I believe the conditional `_CONSTEXPR20_DYNALLOC` is not intended to be transitional - it is expected to be required even once we are able to support MSVC `constexpr` dynamic allocation.  Regarding this change, STL had [questions](https://github.com/microsoft/STL/pull/1369#discussion_r553080412) about the EDG story for WG21-P0784.  I am honestly a little fuzzy on what EDG allows/disallows in terms of `::operator new`, etc. but if it agrees with Clang in allowing it (perhaps indicated by the currently-passing tests), I think that maybe the `#ifdef` should be restored but instead of only allowing it for Clang, we should modify it to instead only _dis_-allow `_CONSTEXPR20_DYNALLOC` for MSVC.  Thoughts?

miscco replied:
I believe the problem is, that MSVC will find that function, because it only replaces `std::allocator::allocate` during contant evaluation. So everywhere else it will find `_Allocate` and it will throw up on it being `constexpr`

Given that we want to use MSVC for constexpr string and constexpr vector with the next release, I do not see any solution other than conditional compilation.

mnatsuhara replied:
Sure, I'm agreeing with the approach of conditional compilation, but am proposing to replace
```c++
#ifdef __clang__
_CONSTEXPR20_DYNALLOC
#endif // __clang__
```
with the following: 
```c++
#if !(defined(__MSVC__)) // or whatever the appropriate way to detect MSVC is
_CONSTEXPR20_DYNALLOC
#endif // __MSVC__
```
This would enable the `_Default_allocate_traits` functions to also be `constexpr` for EDG (since it appears to have no problem with it) and instead only make the functions _not_ `constexpr` for MSVC.  Then, even when we enable MSVC support for `constexpr` dynamic allocation with the next release and `_CONSTEXPR20_DYNALLOC` becomes `constexpr` for MSVC as well, these functions will not.  

If I'm understanding your concern correctly, I think this would address it.

miscco replied:
Yepp, totally misread what you meant :+1:

mnatsuhara replied:
Though now, looking at @StephanTLavavej's [comment](https://github.com/microsoft/STL/pull/1369#discussion_r553080412) more carefully (which I helpfully linked above but evidently did not fully mentally digest), he has some concerns over supporting something for EDG but not MSVC.  

> In general, it's okay to ship a feature enabled for Clang-but-not-MSVC or MSVC-but-not-Clang depending on who's ready. While EDG attempts to match MSVC, it's okay to enable a feature for MSVC but disable it for EDG temporarily (as we did with Concepts). However, customers have found the reverse to be confusing - if a feature is active for EDG but inactive for MSVC, then IntelliSense appears to be happy with the code, yet it emits errors during actual compilation (this happened for <bit>, IIRC).



So it seems like he may want to disable supporting EDG for `constexpr` dynamic allocation until we can also support it for MSVC in the next release.  I'll let him chime in more here given this discussion. 

Note that if we do this, we'll need to modify some of the tests to not expect EDG to accept `constexpr` in some places - as I recall there were some `/BE` tests that were failing before we added `defined(__EDG__)` to the `_CONSTEXPR20_DYNALLOC` macro.

StephanTLavavej replied:
Thanks for catching my mistake!

[@mnatsuhara]
> So, I believe the conditional `_CONSTEXPR20_DYNALLOC` is not intended to be transitional - it is expected to be required even once we are able to support MSVC `constexpr` dynamic allocation.

Aha, the `TRANSITION` comment led to my confusion. 

> So it seems like he may want to disable supporting EDG for constexpr dynamic allocation until we can also support it for MSVC in the next release.

That's correct - I made that change in `yvals_core.h`, in the definition of the feature-test macro. Users are confused when IntelliSense supports features that MSVC doesn't - e.g. DevCom-997767 with `<bit>`.

When MSVC supports `constexpr` dynamic allocation, I believe it would be desirable to use the same codepaths for MSVC and EDG (as EDG IntelliSense is designed to imitate MSVC as much as possible), unless EDG rejects the MSVC approach, in which case we should (1) file an EDG bug and (2) use the Clang strategy for EDG as a workaround. (Perhaps we wouldn't even need to file an EDG bug, if this cannot possibly affect programmer-users, only the library implementation itself.)

> Note that if we do this, we'll need to modify some of the tests to not expect EDG to accept `constexpr` in some places

The tests are currently passing - I believe because they are consistently guarded with the feature-test macro. (The idea is that the feature-test macro should be the only source of truth for whether the feature is available - if it is, code should be able to use the feature without different logic for different compilers, as any such compiler workarounds should be within the product code.)

I'll restore the `#ifdef __clang__` guards to these two `_CONSTEXPR20_DYNALLOC` mentions, preserving their new location, but dropping the `TRANSITION` comments. When we enable MSVC and EDG support in the future, we can rework these guards if we find issues - the feature-test macro won't "light up" automatically, so we'll have a chance to make sure everything is good.

mnatsuhara replied:
Great, thanks for the explanation and fix, @StephanTLavavej!  I think you must be right - when we were hitting `/BE` failures before, it was likely because there were sections that were not correctly guarded by the feature-test macro.  I love seeing those tests green! 😸

miscco commented on stl/inc/xmemory:
```diff
@@ -73,27 +73,18 @@ _INLINE_VAR constexpr size_t _New_alignof = (_STD max)(alignof(_Ty),
-#ifdef __clang__ // TRANSITION, MSVC support for constexpr dynamic allocation
-        _CONSTEXPR20_DYNALLOC
-#endif // __clang__
+    __declspec(allocator) static _CONSTEXPR20_DYNALLOC
```
Same here, this will break once `_CONSTEXPR20_DYNALLOC` turns into `constexpr` for MSVC

StephanTLavavej commented:
It's merged! This will ship in VS 2019 16.10 Preview 1. Thanks for implementing this C++20 feature, and for enduring our exceptionally lengthy code review process, and congratulations on your first microsoft/STL commit! :tada: :rocket: :smile_cat:
